routes/check_api_version.go

```diff
@@ -7,23 +7,35 @@ import (
 )
 
 // CheckAPIVersion checks that the request has a Draupnir-Version header with a
-// version that matches the server's version.
+// version that matches the server's major version, and is equal to or lower
+// than the minor version.
+//
 // If the version doesn't match or the header is missing, it renders a 400 Bad
-// Request
-func CheckAPIVersion(next http.HandlerFunc) http.HandlerFunc {
-	return func(w http.ResponseWriter, r *http.Request) {
-		versions := r.Header["Draupnir-Version"]
-		if len(versions) == 0 {
-			RenderError(w, http.StatusBadRequest, missingApiVersion)
-			return
-		}
+// Request.
+func CheckAPIVersion(serverVersion string) func(http.HandlerFunc) http.HandlerFunc {
+	return func(next http.HandlerFunc) http.HandlerFunc {
+		return func(w http.ResponseWriter, r *http.Request) {
+			versions := r.Header["Draupnir-Version"]
+			if len(versions) == 0 {
+				RenderError(w, http.StatusBadRequest, missingApiVersion)
+				return
+			}
 
-		requestVersion := versions[0]
-		if requestVersion != version.Version {
-			RenderError(w, http.StatusBadRequest, invalidApiVersion(requestVersion))
-			return
-		}
+			major, minor, _, err := version.ParseSemver(serverVersion)
+
+			// If we can't parse our server version then we shouldn't react by rejecting all
+			// requests.
```

Comment:
I reckon it'd be a lot nicer to parse the version on boot and exit if we can't

Reply:
Going to look at using a semver library, might make a bit more sense.

Reply:
If we parse this on boot then it's possible that a standard go build will just produce really useless binaries, which kinda sucks for things like tests etc. It prevents your ability to run `go test file.go`, which I'm not a fan of.

Also tried wrapping all this in a semver library but I'm not sure it warrants the addition of an extra library to do this. It didn't really make this code any nicer!

Want to get this out to cut a new release (for abr) but will revisit this in future.

Reply:
> It prevents your ability to run go test file.go, which I'm not a fan of.

Can you even do this in draupnir at the moment?

Reply:
Yep! It's how I usually invoke tests: ` go test routes/*.go`

Reply:
An individual file doesn't work though, right? Or at least it doesn't for me.

Reply:
You need to specify go files that can be expanded to find the code you want to run. So you couldn't ask it to run tests from a single file without also providing the file that defines the behaviour under test.

Reply:
Anyway, getting back to the point: I'm still quite uneasy about making it possible to build a binary with a bad version that will then just allow any requests, regardless of what version of the client they came from. Obviously it's also bad for the server to reject all requests too - ideally we catch this at compile time, or failing that at boot time.

Reply:
There's an argument that as long as your unit tests pass, your version should be correct, but since we inject a version into the `CheckApiVersion` tests we won't actually catch this.

Reply:
As discussed, added an integration test that verifies we've compiled a binary with a valid semantic version. This should allow testing of the go binary without compiling in a version constant, while still having a check that ensures we don't make a mistake before building a new release.